src/web/demo.rkt

```diff
@@ -443,13 +419,33 @@
 (define (improve req)
   (improve-common
    req
-   (λ (hash formula)
-     (unless (already-computed? hash formula)
-       (semaphore-wait (run-xxx hash formula 'improve (list))))
-
-     (redirect-to (add-prefix (format "~a.~a/graph.html" hash *herbie-commit*)) see-other))
+   (λ (formula)
+    (thread-improve-and-save formula)
+    (redirect-to (add-prefix (format "~a.~a/graph.html" hash *herbie-commit*)) see-other))
    (url main)))
 
+;; Bad name but encapslates shared work of runing the improve command
+;; saving the results to disk
+(define (thread-improve-and-save formula)
```

Comment:
The bad name indicates that it's a bad function. Why combine the improve and save code in one? I'm sure this deduplicates further but the goal of refactoring isn't, like, to compress the code (ZIP works better for that), it's to come up with concepts that make the code easier to understand. I don't understand how the `thread-improve-and-save` concept helps with that.